### PR TITLE
Week 1 Assignment

### PR DESCRIPTION
**Pull request by Akshama**
Code Review.

### CODE REVIEW

File: 1_LongestSubstring.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+Examples:
+
+Given "abcabcbb", the answer is "abc", which the length is 3.
+
+Given "bbbbb", the answer is "b", with the length of 1.
+
+Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
+
+
+//TEBOW IT
```
Comment (omusavi) [inline 1_LongestSubstring.cpp:19]: Great job writing your thoughts here!

File: 1_LongestSubstring.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+*empty string*      0
+abc def             7
+
+
+3. B - BRUTE-FORCE
+
+Check for all substrings starting from each letter of the string and return the longest substring with no repeating characters.
+For n*(n+1)/2 substrings, linear scanning would be done for unique characters with a time complexity of O(n^3).
+4. O - OPTIMISE
+Create a sliding window type of structure that updates the start pointer on each repetition of a character. A substring which is already checked isn't searched again.
```
Comment (omusavi) [inline 1_LongestSubstring.cpp:46]: Good!

File: 1_LongestSubstring.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+    }
+
+    return longest;
+}
+
+//7. T - TEST
+
+bool testlongestSubstring()
```
Comment (omusavi) [inline 1_LongestSubstring.cpp:96]: I think I have a test case that will fail given you implementation.

Can you try the following input string and let me know what happens, and fix it if there are any errors? :)
Input: "abcadef"
Expected output: 6

Comment (Akshama) [reply 1_LongestSubstring.cpp:96]: Thanks a lot for the code review Omeed! :)

I've tried this input string and I'm getting the output as 6 which is expected. 
Here's how my code logic works for this string:
0 1 2 3 4 5 6 
a b c a d e f

Initially, 
curr = 0
longest = 0
visited[x] = -1  for all x from 0 to 256 (essentially, all the ASCII symbols)

**i=0**           
visited[a] = visited[97] = -1
Hence, entering the 'if' block.
curr = 1
visited[a] = 0
longest = 1

**i=1**           
visited[b] = visited[98] = -1
Hence, entering the 'if' block.
curr = 2
visited[b] = 1
longest = 2

**i=2**           
visited[c] = visited[99] = -1
Hence, entering the 'if' block.
curr = 3
visited[c] = 2
longest = 3

**i=3**           
if(visited[a] == -1 || visited[a] < ( i -curr ))
Now, visited[a] = 0 and (i-curr) = (3-3) =0
Hence, condition not satisfied; entering 'else' block.
curr = (i - visited[a]) = (3-0) = 3
visited[a] = i = 3

**i=4**           
visited[d] = visited[100] = -1
Hence, entering the 'if' block.
curr = 4
visited[d] = 4
longest = 4

**i=5**    
visited[e] = visited[101] = -1
Hence, entering the 'if' block.
curr = 5
visited[e] = 5
longest = 5

**i=6**    
visited[f] = visited[102] = -1
Hence, entering the 'if' block.
curr = 6
visited[f] = 6
longest = 6

**for loop ends, returns longest which is 6 in the end, as shown above.**

File: 1_LongestSubstring.cpp (hunk 1 above)
Comment (Akshama) [reply 1_LongestSubstring.cpp:19]: Thankyou!

**Pull request merged**